### PR TITLE
Michaelrfairhurst/declarations8 rule 6-2-3 do not duplicate source code

### PR DESCRIPTION
**Pull request by MichaelRFairhurst**
## Description

This rule is somewhat interesting because 6-2-1 already mostly does things that 6-2-3 does. We can't expressly detect ODR violations, so 6-2-1 looks for repeated names across files. However, there's still room to make 6-2-1 stricter and that's what this query does. Essentially, 6-2-3 gives us permission to flag cases that we considered likely false positives in 6-2-1.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - `RULE-6-2-3`
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/misra/test/rules/RULE-6-2-3/test.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#include <cstdint>
+
+inline int16_t global_redefined = 0;  // NON_COMPLIANT[False negative]
+inline int16_t global_unique = 0;     // COMPLIANT
+inline int16_t global_redeclared = 0; // COMPLIANT
+inline void func_redefined() {}       // NON_COMPLIANT
```
Comment (knewbury01) [inline cpp/misra/test/rules/RULE-6-2-3/test.cpp:6]: is this actually technically compliant?
the rule says inline entities can be defined once per each translation unit, test and test2 are separate translation units? (as long as the definition is the same... however maybe we cant really check that so we should always flag, is that the logic?)

File: cpp/misra/test/rules/RULE-6-2-3/test.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#include <cstdint>
+
+inline int16_t global_redefined = 0;  // NON_COMPLIANT[False negative]
```
Comment (knewbury01) [inline cpp/misra/test/rules/RULE-6-2-3/test.cpp]: minor adjustment to NON_COMPLIANT[FALSE_NEGATIVE]

File: rule_packages/cpp/Declarations8.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          "precision": "very-high",
+          "severity": "error",
+          "short_name": "DuplicateInlineFunctionDefinitions",
+          "tags": [
+            "correctness",
+            "maintainability",
+            "scope/system"
+          ],
+          "implementation_scope": {
+            "description": "This query does not detect duplicated definitions of inline variables.",
```
Comment (knewbury01) [inline rule_packages/cpp/Declarations8.json]: maybe note here that the rest of duplicated entities (noninline, vars etc) will be found in some other rules (list those ones)?

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateInlineFunctionDefinitions.ql [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.cpp.misra
+import codingstandards.cpp.types.Compatible
+
+predicate isInline(FunctionDeclarationEntry d) { d.getDeclaration().isInline() }
+
+predicate interestedInFunctions(FunctionDeclarationEntry f1, FunctionDeclarationEntry f2) {
```
Comment (knewbury01) [inline cpp/misra/src/rules/RULE-6-2-3/DuplicateInlineFunctionDefinitions.ql:33]: classic case, I forget how to solve this , but this predicate seems to take a long time on opencv

**Review by knewbury01 — COMMENTED** [review]
completed partial review, will come back for the rest at a later time, mostly minor comments so far only!

File: rule_packages/cpp/Declarations8.json (hunk 3 above)
Comment (knewbury01) [reply rule_packages/cpp/Declarations8.json]: and maybe also elaborate here on which precise case this rule does detect (fairly obvious from the name but maybe only bc I have also read the implementation)

File: rule_packages/cpp/Declarations8.json [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            "correctness",
+            "maintainability",
+            "scope/system"
+          ]
+        },
+        {
+          "description": "Defining a type with the same fully qualified name in multiple files increases the risk of ODR violations and undefined behavior.",
+          "kind": "problem",
+          "name": "Duplicate type definitions across files",
+          "precision": "very-high",
```
Comment (knewbury01) [inline rule_packages/cpp/Declarations8.json:49]: this query finds pairs that are sometimes related, in openpilot the first result I see is one typedef and then a second def where the second one is only defined (in a header) ifndef the other one, which makes a lot of sense I think? maybe we can just lower the precision to high for this one as a result?

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.cpp.misra
+import codingstandards.cpp.Linkage
+
+class UserTypeDefinition extends TypeDeclarationEntry {
+  UserTypeDefinition() {
+    (isDefinition() or getDeclaration() instanceof TypedefType) and
```
Comment (knewbury01) [inline cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: ok, I am not sure of this myself but wanted to check

should this rule actually include typedefs at all?
[typedefs can be redeclared ](https://en.cppreference.com/cpp/language/typedef)if they refer to the same type so technically just the presence of 2 with the same name is not an ODR? also , if a typedef is a declaration not a defn, maybe incorrect duplicated typedefs (if the types are not the same) is another rule (not sure?) and wouldnt really be ODR? though I am having a hard time finding a resource on that, other than [this](https://en.cppreference.com/cpp/language/definition) which for types just says enum and class types are relevant for ODR

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    not getDeclaration() instanceof ClassTemplateSpecialization and
+    not getDeclaration().getNamespace() instanceof WithinAnonymousNamespace
+  }
+
+  UserType getUserType() { result = getDeclaration() }
+}
+
+from UserTypeDefinition t1, UserTypeDefinition t2
+where
+  not isExcluded(t1, Declarations8Package::duplicateTypeDefinitionsQuery()) and
```
Comment (knewbury01) [inline cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: I think this should have (maybe , double check, bc this then does cut out dups of things in /usr/include/ headers so not totally sure)
```suggestion
  not isExcluded(t1, Declarations8Package::duplicateTypeDefinitionsQuery()) and
  not isExcluded(t2, Declarations8Package::duplicateTypeDefinitionsQuery()) and
```

**Review by knewbury01 — COMMENTED** [review]
another partial review - for duplicate type definitions portion

File: cpp/misra/test/rules/RULE-6-2-3/test.cpp (hunk 1 above)
Comment (MichaelRFairhurst) [reply cpp/misra/test/rules/RULE-6-2-3/test.cpp:6]: It's a good question. Here's my interpretation, please double check me on this!

Under amplification, it says:

> This rule requires that the source code to implement an entity should only appear once ... If the entity is inline, it can be implemented within a header file ... [and included in] multiple translation units

And rationale:

> inline entities can be defined once for each translation unit, but the definitions shall be identical. This is known as the one-definition rule. ... Requiring that the [definition] only appears once reduces the risk of violating the one-definition rule.

Rule 6-2-1 says that the one-definition-rule shall not be violated. My interpretation is that this rule is intended to be stricter than 6-2-1. So multiple definitions of an inline entity in different header files is disallowed, and if that inline entity is used it multiple translation units, it should be implemented in a single header and included in those translation units rather than copied.

wdyt?

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateInlineFunctionDefinitions.ql (hunk 4 above)
Comment (MichaelRFairhurst) [reply cpp/misra/src/rules/RULE-6-2-3/DuplicateInlineFunctionDefinitions.ql:33]: Fixed! Good catch, it didn't have an issue on the dbs I tried. Glad you tried more than I did!

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql (hunk 6 above)
Comment (MichaelRFairhurst) [reply cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: I wondered the same thing!

I eventually figured that typedefs can increase ODR violations like so:

```
// a.h
typedef int32_t my_int;
// b.h
typedef int64_t my_int;
// c.h
void f() {
  my_int x = ...;
  ...
}

// d.cpp
#include "a.h"
#include "c.h"

// e.cpp
#include "b.h"
#include "c.h"
```

I'd be happy to remove typedef checking, because that's a pretty rare kind of scenario.

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql (hunk 7 above)
Comment (MichaelRFairhurst) [reply cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: good catch, I agree.

File: rule_packages/cpp/Declarations8.json (hunk 3 above)
Comment (MichaelRFairhurst) [reply rule_packages/cpp/Declarations8.json]: I elaborated, feedback welcome!

File: rule_packages/cpp/Declarations8.json (hunk 5 above)
Comment (MichaelRFairhurst) [reply rule_packages/cpp/Declarations8.json:49]: This an interesting case. The `#ifndef` pattern here is unusual. It's checking if `sighandler_t` is defined as a preprocessor symbol, not as a typedef...

`sighandler_t` is supposed to be exposed by the `#include <csignal>`. It looks like maybe this is some weird portability hack, or attempted portability hack.

Perhaps some runtimes don't expose `sighandler_t`, and so this file redefines it. In this case, when `sighandler_t` is already defined as a `typedef`, and the types are compatible, then the repeated declaration still compiles. But maybe the author was worried some runtimes might expose `sighandler_t` as a macro, e.g.:

```
typedef void (*__sighandler_t)(int sig);
#define sighandler_t __sighandler_t
```

But even in this case, the `#ifndef` seems unnecessary because expanding `sighandler_t` to `__sighandler_t` would still expand to a valid typedef (`typedef void (*__sighandler_t)(int sig);`).

Asking copilot for thoughts, it says:

> The #ifndef guard is effectively a no-op in every realistic scenario. It doesn't hurt anything, but it doesn't actually protect against anything either. It's cargo-cult defensiveness.

I don't know if I'm missing something special here.

I'd say at the end of the day:
- The `#ifndef` is probably not doing anything -- especially since it's only used for one of the two cases
- The two definitions should probably be merged into one `csignal_compat.h` type file
- If the `#ifndef` hack is required for portability, that's all the more reason to unify the two `sighandler_t` definitions

But, uh, yeah, this case is weird...

File: cpp/misra/test/rules/RULE-6-2-3/test.cpp (hunk 1 above)
Comment (knewbury01) [reply cpp/misra/test/rules/RULE-6-2-3/test.cpp:6]: sure, I am happy to leave it with that (stricter) interpretation

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql (hunk 6 above)
Comment (knewbury01) [reply cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: hm , yeah I think still removing typedefs would make sense to me (and make a note in the package json for that it doesnt cover them)

right now there are 200 results on open pilot for them, many seem to be fine redeclarations with the same type, and many more are matches of unnamed struct and unnamed struct (for 2 diff typedef structs) which these result categories are not too useful atm without more work to maybe separate them out further

File: rule_packages/cpp/Declarations8.json (hunk 5 above)
Comment (knewbury01) [reply rule_packages/cpp/Declarations8.json:49]: will be handled by the comment below

File: cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql (hunk 6 above)
Comment (MichaelRFairhurst) [reply cpp/misra/src/rules/RULE-6-2-3/DuplicateTypeDefinitions.ql]: Done!

**Review by knewbury01 — APPROVED** [review]
Thanks for addressing all of the comments! lgtm!
